### PR TITLE
Improve emoji and mention matching

### PR DESCRIPTION
**Pull request by yardenshoham**
Prioritize matches that start with the given text, then matches that contain the given text.

I wanted to add a heart emoji on a pull request comment so I started writing `:`, `h`, `e`, `a`, `r` (at this point I still couldn't find the heart), `t`... The heart was not on the list, that's weird - it feels like I made a typo or a mistake. This fixes that.

This also feels more like GitHub's emoji auto-complete.

# Before
image

# After
image

### CODE REVIEW

**Comment by silverwind** [comment]
This kind of screams for extraction of the matcher to a function and a unit test to me.

**Comment by silverwind** [comment]
Maybe, if one searches hard enough, there might be npm packages for this job 😆.

**Comment by yardenshoham** [comment]
I tried with `string-similarity` but found that what I came up with provides a better experience

**Comment by silverwind** [comment]
Working on a better version, will push here.

**Comment by silverwind** [comment]
New matcher functions up, they are much more smart now and will also match aliases of emojis:

Screenshot 2023-04-21 at 20 20 40
Screenshot 2023-04-21 at 20 24 06

**Comment by silverwind** [comment]
The data at `/assets/emoji.json` is generated, by I think `build/generate-emoji.go`, but running that only does minor changes to `modules/emoji/emoji_data.go`. I guess this generation is broken, I also see no `make` target. Something we should eventually check out.

File: web_src/js/utils/match.test.js [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+    'hankey',
+    'poodle',
+    'spoon',
+    'bowl_with_spoon',
+  ]);
+});
+
+test('matchMention', () => {
+  expect(matchMention('')).toEqual(window.config.tributeValues.slice(0, 6));
+  expect(matchMention('user4')).toEqual([window.config.tributeValues[3]]);
```
Comment (delvh) [inline web_src/js/utils/match.test.js:46]: Wait, we got rid of tribute but still call it tribute?

**Comment by jolheiser** [comment]
> Related: The data at `assets/emoji.json` is generated, by I think `build/generate-emoji.go`, but running that only does minor changes to `modules/emoji/emoji_data.go`. I guess this generation is broken, I also see no `make` target. Something we should eventually check out because the current emoji data is very outdated.

I believe it's a manual run, what part is broken? It grabs the JSON from `https://raw.githubusercontent.com/github/gemoji/master/db/emoji.json` and filters out any `<=` unicode version 14 currently.

The last update was likely https://github.com/go-gitea/gitea/pull/22342

**Review by delvh — APPROVED** [review]

**Comment by silverwind** [comment]
> > Related: The data at `assets/emoji.json` is generated, by I think `build/generate-emoji.go`, but running that only does minor changes to `modules/emoji/emoji_data.go`. I guess this generation is broken, I also see no `make` target. Something we should eventually check out because the current emoji data is very outdated.
> 
> I believe it's a manual run, what part is broken? It grabs the JSON from `https://raw.githubusercontent.com/github/gemoji/master/db/emoji.json` and filters out any `<=` unicode version 14 currently.
> 
> The last update was likely #22342

The `assets/emoji.json` was not being updated with new aliases and such.

File: web_src/js/utils/match.test.js (hunk 1 above)
Comment (silverwind) [reply web_src/js/utils/match.test.js:46]: Tribute is still in there to support easymde editor. the plan is to rip it out maybe for 1.21 or later.

**Comment by jolheiser** [comment]
> > > Related: The data at `assets/emoji.json` is generated, by I think `build/generate-emoji.go`, but running that only does minor changes to `modules/emoji/emoji_data.go`. I guess this generation is broken, I also see no `make` target. Something we should eventually check out because the current emoji data is very outdated.
> > 
> > 
> > I believe it's a manual run, what part is broken? It grabs the JSON from `https://raw.githubusercontent.com/github/gemoji/master/db/emoji.json` and filters out any `<=` unicode version 14 currently.
> > The last update was likely #22342
> 
> The `assets/emoji.json` was not being updated with new aliases and such.

Strange, it should have been writing it https://github.com/go-gitea/gitea/blob/65fe0fb22cfb264f0b756065d0c3ce7a17d7e55b/build/generate-emoji.go#L157
Assuming it was ran from the project root.

**Comment by silverwind** [comment]
Or more specifically, I'm missing the "love" alias to "heart". This works on GitHub as an alias, but I see it's a "tag" in gemoji data, not and "alias" which we don't extract.

https://github.com/github/gemoji/blob/5476a66d2794e0d1551b1f96e449afc72e9f7bec/db/emoji.json#L1845

Screenshot 2023-04-21 at 20 48 11

File: web_src/js/utils/match.test.js (hunk 1 above)
Comment (delvh) [reply web_src/js/utils/match.test.js:46]: Ah right, there was something.

**Comment by jolheiser** [comment]
Ah okay, so we would potentially need to pull out tags and add them to some sort of weight system as well

**Comment by silverwind** [comment]
> Ah okay, so we would potentially need to pull out tags and add them to some sort of weight system as well

There is no real weight system, but earlier entries in the aliases have higher weight, so I guess we can likely just append `tags` to `aliases`.

**Comment by silverwind** [comment]
>  I guess we can likely just append tags to aliases.

Probably best to just preserve `tags` into the JSON. Altering the JS to use it is trivial.

**Comment by mrsdizzie** [comment]
I'm not sure! It looks like tags are perhaps newer or just much less common and weren't considered initially? From what I could see in a quick look at upstream commit comments the reason for tags is that aliases are supposed to be unique and some things make more sense as tags which can be shared across multiple emoji? Based in that I think you could add tags to aliases but only for this JavaScript drop-down and not in the go data? Tags aren't unique so if somebody wrote :love: and expected it to render the same always it would need to map to a unique Alias but multiple emoji could have the tag "love"

**Comment by silverwind** [comment]
Checking GitHub HTML, it seems the tag is just a concatenated into `data-text` which is use for the matching, e.g. `data-text="heart love"` for this emoji. Still I suggest we just output `tags` into the json file for the JS to consume independently.

**Comment by silverwind** [comment]
@mrsdizzie would there be an issue if the backend does not know about the tags, but frontend does? The output will still be `:heart:` in the frontend, so I think it'll be fine.

**Comment by mrsdizzie** [comment]
Yea that should be fine I assume it would keep working as-is. The drop down should insert the actual emoji character and not an Alias so any backend processing would only ever see that. I don't think there is any technical connection between front end and backend dataset other than they are generated from the same place and expected to have similar results. Adding additional tag results to JavaScript search should be fine and seems like what they are intended for

**Comment by silverwind** [comment]
Thanks, clear. I can confirm this new code will only ever return the first `alias` to the backend and backend can render all aliases (previous tribute code could also return non-first aliases), so it'll be fine.

**Comment by yardenshoham** [comment]
I tested. This feels a bit weird:
image

But as long as `:heart` shows the heart (which it does) it's an LGTM from my side.
image

**Comment by silverwind** [comment]
That exclamation appearing there is a bug I'd say. It's matching because it has a alias `heavy_exclamation_mark` but we need to downrank such secondary matches. I'll think about something.

**Comment by yardenshoham** [comment]
I feel like my initial approach of 1. starts-with, 2. includes was pretty good. You could do 1. starts-with, 2. includes, 3. includes-alias or 1. starts-with, 3. includes-alias, includes

**Comment by silverwind** [comment]
Currently every match has a "score" based on match index, the lower it is, the earlier it will be ranked. I think we could just add +1 to score for 1st alias, +2 for second alias etc.

**Comment by silverwind** [comment]
Implemented above, please try again.

**Comment by yardenshoham** [comment]
Yes, that's way better

**Comment by silverwind** [comment]
Feel free to add more tests if you like. Just keep in mind that we will probably alter them on the next emoji data update.

**Comment by yardenshoham** [comment]
Let's land it

**Comment by silverwind** [comment]
Yes, need a review for that 😆

**Review by silverwind — APPROVED** [review]

**Pull request merged**